### PR TITLE
Bump cluster-api to v1.0.2

### PR DESCRIPTION
**Pull request by thedadams**
In addition to bumping the CAPI dependency, this PR adds conversion webhook
configuration to CAPI CRDs on startup 

Since the CAPI CRD versions have been bumped to v1beta1, we need to provide a
way for the old objects with version v1alpha4 to be upgraded to the new version.
The convert webhook is already running in rancher-webhook, so the conversion
configuration is added to the CRDs on startup.

This conversion configuration cannot be added to the CRD ahead of time because
the CABundle is not known. We rely on the `needacert` controller to recognize
that the conversion configuration needs the CABundle, and it will be added when
the CRD is updated. If the conversion webhook configuration was added
beforehand, then Rancher startup would hang as it tries to watch the CAPI
objects. It would not be able to watch the CAPI objects because doing so would
call the conversion webhook which would fail because of an invalid cert (we
don't know the certificate information). Therefore, this migration is necessary.

NOTE: This PR will not be complete until the [rancher-webhook](https://github.com/rancher/webhook/pull/64) PR is merged, a new RC tag created there, and the rancher-webhook version bumped here.

Issue:
https://github.com/rancher/rancher/issues/35489

### CODE REVIEW

File: pkg/api/norman/customization/monitor/prometheus_manager.go [hunk 1]
```diff
@@ -126,7 +126,7 @@ func (q *Queries) Do(querys []*PrometheusQuery) (map[string][]*TimeSeries, error
 }
 
 func (q *Queries) GetLabelValues(labelName string) ([]string, error) {
-	value, _, err := q.api.LabelValues(q.ctx, labelName, []string{}, time.Now().Add(-time.Hour), time.Now())
+	value, _, err := q.api.LabelValues(q.ctx, labelName, nil, time.Now().Add(-time.Hour), time.Now())
```
Comment (snasovich) [inline pkg/api/norman/customization/monitor/prometheus_manager.go:129]: This change looks irrelevant to the scope of the PR to me.

File: pkg/rancher/migrations.go [hunk 2]
```diff
@@ -228,3 +230,38 @@ func applyProjectConditionForNamespaceAssignment(label string, condition conditi
...
+		}
+
+		crd.Spec.Conversion = &apiextv1.CustomResourceConversion{
+			Strategy: apiextv1.WebhookConverter,
+			Webhook: &apiextv1.WebhookConversion{
+				ClientConfig: &apiextv1.WebhookClientConfig{
+					Service: &apiextv1.ServiceReference{
+						Namespace: "cattle-system",
+						Name:      "webhook-service",
+						Path:      &[]string{"/convert"}[0],
```
Comment (snasovich) [inline pkg/rancher/migrations.go:253]: This is both ugly and neat at the same time :)

**Review by snasovich — CHANGES_REQUESTED** [review]
One general comment and one question. Otherwise LGTM.
However, the build is failing due to go modules verification issue. It doesn't look to me that it's caused by pending webhook bump, so please take a look.

File: pkg/api/norman/customization/monitor/prometheus_manager.go (hunk 1 above)
Comment (thedadams) [reply pkg/api/norman/customization/monitor/prometheus_manager.go:129]: Ah yes. This is leftover from my original switch, and should not be changed. Will update.

File: pkg/provisioningv2/capi/capi.go [hunk 3]
```diff
@@ -13,10 +13,15 @@ import (
...
 	clusterv1alpha3 "sigs.k8s.io/cluster-api/api/v1alpha3"
 	clusterv1alpha4 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	clusterv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/cluster-api/controllers"
 	"sigs.k8s.io/cluster-api/controllers/remote"
 	addonsv1alpha3 "sigs.k8s.io/cluster-api/exp/addons/api/v1alpha3"
-	expv1alpha3 "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	addonsv1alpha4 "sigs.k8s.io/cluster-api/exp/addons/api/v1alpha4"
+	addonsv1beta1 "sigs.k8s.io/cluster-api/exp/addons/api/v1beta1"
+	expvv1alpha3 "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
```
Comment (jakefhyde) [inline pkg/provisioningv2/capi/capi.go:22]: Should this be `expv1alpha3`? Same question for `alpha4` and `beta1`.

Comment (jakefhyde) [inline pkg/provisioningv2/capi/capi.go:22]: Should this be `expv1alpha3`? Same question for `alpha4` and `beta1`.

**Review by jakefhyde — CHANGES_REQUESTED** [review]

File: pkg/provisioningv2/capi/capi.go (hunk 3 above)
Comment (thedadams) [reply pkg/provisioningv2/capi/capi.go]: Fixed.

File: pkg/api/norman/customization/monitor/prometheus_manager.go (hunk 1 above)
Comment (thedadams) [reply pkg/api/norman/customization/monitor/prometheus_manager.go]: Fixed.

**Review by jakefhyde — APPROVED** [review]

File: pkg/provisioningv2/capi/capi.go [hunk 4]
```diff
@@ -26,8 +31,13 @@ func init() {
 	_ = clientgoscheme.AddToScheme(schemes.All)
 	_ = clusterv1alpha3.AddToScheme(schemes.All)
 	_ = clusterv1alpha4.AddToScheme(schemes.All)
+	_ = clusterv1beta1.AddToScheme(schemes.All)
 	_ = expv1alpha3.AddToScheme(schemes.All)
+	_ = expv1alpha4.AddToScheme(schemes.All)
```
Comment (kinarashah) [inline pkg/provisioningv2/capi/capi.go:36]: How do we identify we need exp v1alpha4? I looked at `capi-webhooks.yaml` and all entries are using v1beta1.

Comment (thedadams) [reply pkg/provisioningv2/capi/capi.go:36]: I am not entirely sure how we would verify which schemes we need, especially with all the upgrade scenarios.

Comment (kinarashah) [reply pkg/provisioningv2/capi/capi.go:36]: Oh okay, so we need to add all of them when bumping capi 👍🏼

**Review by kinarashah — DISMISSED** [review]

**Comment by thedadams** [comment]
Rancher webhook is now bumped here.

File: package/Dockerfile [hunk 5]
```diff
@@ -46,7 +46,7 @@ ENV CATTLE_SYSTEM_AGENT_UPGRADE_IMAGE rancher/system-agent:${CATTLE_SYSTEM_AGENT
 
 # System charts minimal version
 ENV CATTLE_FLEET_MIN_VERSION=100.0.2+up0.3.8
-ENV CATTLE_RANCHER_WEBHOOK_MIN_VERSION=1.0.2+up0.2.2
+ENV CATTLE_RANCHER_WEBHOOK_MIN_VERSION=1.0.3+up0.2.4-rc2
```
Comment (snasovich) [inline package/Dockerfile:49]: Doesn't prerelease go before metadata usually? 
https://regex101.com/r/1vI80U/1 - using regex from https://semver.org/

**Review by snasovich — COMMENTED** [review]
Have one question, otherwise LGTM

File: package/Dockerfile (hunk 5 above)
Comment (thedadams) [reply package/Dockerfile:49]: This is the way rancher-charts works.

Comment (thedadams) [reply package/Dockerfile:49]: We are releasing a new version of the chart (1.0.3) and the image contained in this version is 0.2.4-rc2. Once the chart is released as part of rancher/charts release, then the version of the image will be 0.2.4

**Review by jiaqiluo — APPROVED** [review]
LGTM 💯
